File: src/main/java/me/andre111/d20server/service/ModuleService.java
package me.andre111.d20server.service;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import com.google.gson.Gson;

import me.andre111.d20common.model.def.Definitions;
import me.andre111.d20common.model.def.EntityDefinition;
import me.andre111.d20common.model.def.EntityDefinition.ExtensionPoint;
import me.andre111.d20common.model.def.ExtensionDefinition;
import me.andre111.d20common.util.Utils;
import me.andre111.d20server.server.HttpServerHandler;

public abstract class ModuleService {
	private static List<Module> modules = new ArrayList<>();
	
	public static void init() {
		init(new File("./modules/"));
	}
	public static void init(File modulesDir) {
		// load modules
		//modules.add(new Module("base", new File(modulesDir, "/base/"))); //TODO: allow full custom modules (including removing base?)
		modules.add(new Module("core", new File(modulesDir, "../core/"))); //TODO: allow full custom modules (including removing base?)
		
		for(File file : modulesDir.listFiles()) {
			if(file.isDirectory() && !file.getName().equals("base")) {
				modules.add(new Module(file.getName(), file));
			}
		}
		
		load();
	}
	
	public static Stream<Module> enabledModules() {
		return modules.stream().filter(module -> module.isEnabled());
	}
	
	private static void load() {
		// load definitions
		loadDefinitions();
		Definitions.init();
		HttpServerHandler.initModules();
	}
	
	
	private static void loadDefinitions() {
		// load entity definitions
		for(Map.Entry<String, File> e : getFilesIn("entities").entrySet()) {
			if(e.getValue().isFile() && e.getKey().endsWith(".json")) {
				// read basic entity definition
				String type = e.getKey().substring(0, e.getKey().length()-5);
				EntityDefinition entityDefinition = Utils.readJson(e.getValue(), EntityDefinition.class);
				System.out.println("Entity: "+type);
				
				// read extension definitions
				for(ExtensionPoint extensionPoint : entityDefinition.extensionPoints()) {
					for(Map.Entry<String, File> ee : getFilesIn("entities/extensions/"+type+"/"+extensionPoint.name()+"/").entrySet()) {
						if(ee.getValue().isFile() && ee.getKey().endsWith(".json")) {
							String extensionName = ee.getKey().substring(0, ee.getKey().length()-5);
							ExtensionDefinition extensionDefinition = Utils.readJson(ee.getValue(), ExtensionDefinition.class);
							System.out.println("    "+extensionPoint.name()+": "+extensionName);
							
							// add extension definition
							extensionPoint.extensionDefinitions().put(extensionName, extensionDefinition);
						}
					}
				}
				
				// add entity definition
				Definitions.addEntityDefinition(type, entityDefinition);
			}
		}
	}
	
	private static Map<String, File> getFilesIn(String path) {
		Map<String, File> files = new HashMap<>();
		// reject paths that could lead to leaving the module directory
		if(path.contains("..")) return files;

		enabledModules().forEach(module -> {
			File moduleDir = new File(module.directory, path);
			if(moduleDir.exists()) {
				for(File file : moduleDir.listFiles()) {
					files.put(file.getName(), file);
				}
			}
		});
		
		return files;
	}
	
	public static final class Module {
		private final transient String identifier;
		private final transient File directory;
		private final ModuleDefinition definition;
		
		public Module(String identifier, File directory) {
			this.identifier = identifier;
			this.directory = directory;
			
			if(!identifier.equals("core")) {
				try(InputStreamReader reader = new InputStreamReader(new FileInputStream(new File(directory, "module.json")), "UTF-8")) {
					this.definition = (new Gson()).fromJson(reader, ModuleDefinition.class);
				} catch(IOException e) {
					throw new RuntimeException("Invalid module: "+identifier, e);
				}
			} else {
				this.definition = new ModuleDefinition("core", "0", "temporary hack", false, new ArrayList<>());
			}
		}
		
		public String getIdentifier() {
			return this.identifier;
		}
		
		public ModuleDefinition getDefinition() {
			return this.definition;
		}
		
		public File getDirectory() {
			return this.directory;
		}
		
		public boolean isEnabled() {
			return true; //TODO: implement
		}
	}
	public static final record ModuleDefinition(String name, String version, String description, boolean disableable, List<String> libraries) {
	}
}

File: core/server/service/module-service.js
import path from 'path';
import fs from 'fs';

import { readJsonFile } from '../util/fileutil.js';
import { Definitions } from '../../common/common.js';
import { setDefinitions } from '../../common/definitions.js';
import { MessageService } from './message-service.js';
import { ModuleDefinitions } from '../../common/messages.js';
import { UserService } from './user-service.js';
import { Role } from '../../common/constants.js';
import { CONFIG } from '../config.js';
import { I18N } from '../../common/util/i18n.js';

class Module {
    identifier;
    directory;
    definition;
    enabled;

    constructor(identifier, directory) {
        this.identifier = identifier;
        this.directory = directory;

        this.definition = readJsonFile(path.join(directory, 'module.json'));
        this.enabled = !CONFIG.get().disabledModules.includes(identifier);
    }

    getIdentifier() {
        return this.identifier;
    }

    getDirectory() {
        return this.directory;
    }

    getDefinition() {
        return this.definition;
    }

    isEnabled() {
        return this.enabled;
    }
}

var modules = [];
export class ModuleService {
    static async init() {
        // "init" config
        //TODO: move this to an actual "config option definition" system
        CONFIG.get().disabledModules = CONFIG.get().disabledModules ?? [];
        CONFIG.get().language = CONFIG.get().language ?? 'de_DE';
        CONFIG.get().gmLockout = CONFIG.get().gmLockout ?? false;
        CONFIG.get().motd = CONFIG.get().motd ?? '';

        // scan for modules
        fs.readdirSync(path.join(path.resolve(), '/modules/')).forEach(file => {
            const directory = path.join(path.resolve(), '/modules/'+file+'/');
            if(fs.statSync(directory).isDirectory()) {
                // check for module definition and add module
                if(fs.existsSync(path.join(directory, 'module.json'))) {
                    modules.push(new Module(file, directory));
                }
            }
        });

        // load i18n files
        ModuleService.loadI18N();

        // load definitions
        ModuleService.loadDefinitions();

        // load code
        for(const module of modules) {
            if(module.isEnabled()) {
                const commonJSFile = path.join(module.getDirectory(), '/common/module.js');
                if(fs.existsSync(commonJSFile)) {
                    await import('../../../modules/'+module.getIdentifier()+'/common/module.js');
                }

                const serverJSFile = path.join(module.getDirectory(), '/server/module.js');
                if(fs.existsSync(serverJSFile)) {
                    await import('../../../modules/'+module.getIdentifier()+'/server/module.js');
                }
            }
        }
    }

    static loadI18N() {
        const loadLangFile = file => { if(fs.existsSync(file)) I18N.mergeObject(readJsonFile(file)); };
        const lang = CONFIG.get().language;

        loadLangFile('./core/i18n/'+lang+'.json');
        ModuleService.forEnabledModules(module => loadLangFile(path.join(module.getDirectory(), 'i18n/'+lang+'.json')));
    }

    static loadDefinitions() {
        const definitions = new Definitions();

        for(const [name, file] of Object.entries(ModuleService.getFilesIn('/entities/'))) {
            if(name.endsWith('.json')) {
                // read basic entity definition
                const type = name.substring(0, name.length-5);
                const entityDefinition = readJsonFile(file);
                console.log(`Entity: ${type}`);

                // read extension definitions
                for(const extensionPoint of entityDefinition.extensionPoints) {
                    extensionPoint.extensionDefinitions = {};
                    for(const [ename, efile] of Object.entries(ModuleService.getFilesIn('/entities/extensions/'+type+'/'+extensionPoint.name+'/'))) {
                        if(ename.endsWith('.json')) {
                            const extensionName = ename.substring(0, ename.length-5);
                            const extensionDefinition = readJsonFile(efile);
                            console.log('    '+extensionPoint.name+': '+extensionName);

                            // add extension definition
                            extensionPoint.extensionDefinitions[extensionName] = extensionDefinition;
                        }
                    }
                }

                // add entity definition
                definitions.addEntityDefinition(type, entityDefinition);
            }
        }

        setDefinitions(definitions);
    }

    static forEnabledModules(func) {
        for(const module of modules) {
            if(module.isEnabled()) {
                func(module);
            }
        }
    }

    static getFilesIn(dirPath) {
        var files = {};
        const fileScanner = (dir) => {
            if(fs.existsSync(dir)) {
                fs.readdirSync(dir).forEach(file => {
                    files[file] = path.join(dir, file);
                });
            }
        };

        // Scan core + enabled modules
        fileScanner(path.join(path.resolve(), '/core' + dirPath));
        ModuleService.forEnabledModules(module => {
            const dir = path.join(module.getDirectory(), dirPath);
            fileScanner(dir);
        });
        return files;
    }

    static broadcastModuleDefinitions() {
        UserService.forEach(profile => {
            if(profile.getRole() == Role.GM) this.sendModuleDefinitions(profile);
        });
    }

    static sendModuleDefinitions(profile) {
        var moduleDefinitions = {};
        for(const module of modules) {
            moduleDefinitions[module.identifier] = module.definition;
        }

        MessageService.send(new ModuleDefinitions(moduleDefinitions, CONFIG.get().disabledModules), profile);
    }

    static toggleModule(identifier, disabled) {
        const index = CONFIG.get().disabledModules.indexOf(identifier);
        if(disabled) {
            if(index == -1) CONFIG.get().disabledModules.push(identifier);
        } else {
            if(index >= 0) CONFIG.get().disabledModules.splice(index, 1);
        }
        CONFIG.save();
        ModuleService.broadcastModuleDefinitions();
    }
}
